debug/slides_source/debug.tex

```diff
@@ -0,0 +1,382 @@
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        }
+\begin{lstlisting}
+Backtrace when the node is created:
+  File "test.py", line 7, in <module>
+    z = z * y
+
+\end{lstlisting}
+\end{frame}
+
+\subsection{DebugMode}
+\begin{frame}[fragile]
+  \frametitle{DebugMode}
+    Checks and double-checks everything, extremely slow
+\begin{itemize}
+\item Compare Python, C and GPU implementations.
+\item Compare before and after each optimization.
+\item By default, raise an error on NaN.
+\item Sensitive: so frequently reported errors are ok.
+\end{itemize}
+\end{frame}
+
+\subsection{Breakpoint}
+\begin{frame}[fragile]
+  \frametitle{Breakpoint Op} Identity-like op with the side effect of
+  enforcing a conditional breakpoint, inside a theano function, based
+  on a symbolic scalar condition.
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Breakpoint Op: optimization}
+  Theano optimization don't know it.
+\begin{itemize}
+  \item Could slow down computation, but not too much.
+  \item Could disable stability optimization
+  \item It is moved to the GPU (it do not force transfer)
+\end{itemize}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Breakpoint Op: example}
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        basicstyle=\footnotesize,
+        xleftmargin=-1em
+        }
+\begin{lstlisting}
+import theano
+import theano.tensor as T
+from theano.tests.breakpoint import PdbBreakpoint
+input, target = T.fvectors('xy')
+
+mse = (input - target) ** 2
+
+# Conditional breakpoint to be activated if the total MSE is higher
+# than 100. The breakpoint will monitor the inputs, targets as well
+# as the individual error values
+breakpointOp = PdbBreakpoint(``MSE too high'')
+condition = T.gt(mse.sum(), 100)
+mse, monitored_input, monitored_target = breakpointOp(
+    condition, mse, input, target)
+\end{lstlisting}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Breakpoint Op: example}
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        basicstyle=\footnotesize,
+        xleftmargin=-1em
+        }
+\begin{lstlisting}
+# Compile the theano function
+fct = theano.function([input, target], mse)
+
+# Use the function
+print fct([10, 0], [10, 5]) # Will NOT activate the breakpoint
+print fct([0, 0], [10, 5]) # Will activate the breakpoint
+\end{lstlisting}
+
+\end{frame}
+
+\subsection{PrintOp}
+\begin{frame}[fragile]
+  \frametitle{PrintOp}
+Allow to print during execution
+\lstset{language=Python,
+        commentstyle=\itshape\color{blue},
+        stringstyle=\color{violet},
+        basicstyle=\footnotesize,
+        xleftmargin=-1em
+        }
+\begin{lstlisting}
+import theano
+x = theano.tensor.vector()
+o = theano.printing.Print(``a message'')(x)
+f = theano.function([x], o)
+d = f([3, 4])
+a message  __str__ = [ 3.  4.]
+\end{lstlisting}
+\end{frame}
```

Comment:
Both the print and breakpoint ops require that at least one output be used to compute at least one of the outputs of the theano function. It might not be 100% necessary to mention here but it would be nice if the information would be present in this presentation if people want to refer to it. 

If they try to use those during the exercices, they might have a surprise when/if it does not work as they think should